[DRAFT] Resource discovery and conflict resolution proposal for app workloads (#66)
margo/specification-enhancements: proposals/profile-discovery-and-resolution/01-capability-framework.md

### Problem 1 — Device-specific values are unknown at authoring time

A GPU device path, an available network port, or a vault secret path are only known by the platform at deployment time. Today, end-user persona must manually discover and supply these values. If the value changes or becomes invalid, the deployment fails at runtime with unclear errors.

### Problem 2 — Resource conflicts are only detected at runtime

Two applications deployed to the same device both request port `8080` or hostname `app.example.com`. Only one can succeed. The conflict surfaces at runtime — through infrastructure errors, not through the Margo management plane.

### Problem 3 — Discovery logic is scattered and inconsistent

There is no standard mechanism for applications to declare "I need a GPU with 8 GiB VRAM" or "I need an available network port" or "I need a secret from the vault." Each implementation invents its own approach. No cross-implementation conflict detection is possible.

---

## Framework Concepts

> ### Extending ProfileDefinition

Reviewer: I think something may have been lost in the re-write/edits. In the original idea, there was a mechanism for the device to indicate how many allocatable units were available (e.g., GPUs) so the WFM could determine if the device had that resource free. 

Between the rewrite of this SUP and the creation of the other SUP, I'm not seeing that anymore.

Author: That mechanism had a fundamental flaw. The goal was to let WFMs run a quick pre-check against the device profile state, but that state is essentially a cached, stale entity. Given how fast local device activity changes and how often they are reported to WFM, the pre-check wouldn't be accurate.

Author: It was just an enhancement, it's safer to leave it out for now and add it later if we find a clear need.

Reviewer: >Given how fast local device activity changes and how often they are reported to WFM, the pre-check wouldn't be accurate.

Why do you feel this will be changing quickly?

I see application deployments occurring very infrequently, except for maybe the first time a device is provisioned and set up with all the desired applications.  Customers are not going to be deploying new apps or updating deployed apps very often. This is not something that is going to be happening multiple times a day or week.

> As per [`sup_bisect_device_manifest`] :

Reviewer: Did you intend for the "[`sup_bisect_device_manifest`]" part to be a link to the other SUP?

Author: Yeah. I'll push it. Thanks.

**Example:**

```yaml
"requestSchema": {
  "type": "object",
  "properties": {
    "minVramGiB": {
      "type": "number"
    },
    "architecture": {
      "type": "string"
    }
  }
}
```

#### outputSchema

The values the platform will generate upon successful resolution. These values are unknown at authoring time and only exist after evaluation. For guard-mode discoveries, this SHOULD be an empty object `{}`.

**Example:**

```yaml
"outputSchema": {
  "type": "object",
  "properties": {
    "devicePath": {
      "type": "string"
    },
    "vramGiB": {
      "type": "number"
    }
  }
}
```

#### failureCodes

A set of named failure outcomes the platform may return if resolution cannot succeed. Each code is machine-readable and actionable.

**Example:**

```yaml
"failureCodes": [
  "NoGPUAvailable",
  "InsufficientVRAM"
]
```

## Discovery vs. Allocation Policy

**Important distinction:** Discovery is inherently a declarative inquiry, not a guarantee of reservation. The framework allows an application to declare its resource requirements; the platform then evaluates this declaration against the live infrastructure state.

> The core specification defines only the resolution contract (inputs, outputs, and failure modes). It deliberately decouples resource resolution from lifecycle enforcement, and how the resolution is done are implementation defined, for example:

Reviewer: ```suggestion
The core specification defines only the resolution contract (inputs, outputs, and failure modes). It deliberately decouples resource resolution from lifecycle enforcement, and how the resolution is done is an implementation detail, for example.
```

* Transient Resolution: A storage class discovery may simply return a metadata name without locking the underlying pool.
* Implicit Reservation: A network port discovery may mark the resolved port as occupied immediately upon evaluation.
* Exclusive Enforcement: A GPU discovery may trigger strict hardware isolation or exclusive runtime access.

Because resource types have fundamentally different lifecycles, the exact enforcement policy is deferred to the underlying platform implementation.

> Note: Need some feedback/guidance/brainstorming on whether the resource resolution should be defined by Margo, otherwise there will not be a deterministic behaviour.

Reviewer: What, specifically, do you mean by "resource resolution"? 

Based on the text above this note, I'm interpreting this as how the client handles allocating the resources (if they are allocatable). I don't think Margo should be defining any implementation details. The way a device supplier chooses to implement these things is inside the black box. Margo should focus on how the required information is communicated to all parties needing the information. 

However, since you mentioned not defining this makes things non-deterministic, maybe I'm misinterpreting.

Author: By "resource resolution," I mean the exact point where the platform evaluates the application's declared requirements against the live infrastructure state and decides it's a match.

Reviewer: This should be the device's responsibility. The WFM should just be responsible for passing the information the device requires via the application deployment YAML. The device should handle everything that's needed and report any errors it has.

Author: You are completely right that the internal implementation details of how a vendor manages its resources should remain a black box. However, the text currently doesn't discuss the the side effects of discovery and leaves them entirely up to the implementation (e.g., Transient Resolution vs. Implicit Reservation), we introduce a silent race condition that breaks deterministic behavior.

For Example:
* Imagine App 1 and App 2 are pushed to a device at the exact same time, and both require Port 8080.
* If Vendor A implements Implicit Reservation: App 1's discovery phase runs first and immediately marks Port 8080 as occupied. A millisecond later, App 2's discovery phase runs, sees it's occupied, and safely fails.  (This is deterministic behavior).
* If Vendor B implements Transient Resolution: App 1's discovery check runs and returns true without locking. Before App 1 can fully deploy and bind the port, App 2's discovery check runs, sees the port is still technically free, and also returns true. Both pass discovery, both attempt to deploy, and one crashes randomly at runtime. (Non-deterministic behavior). 

The Discovery mechanism fails here. Vendors could fix this race condition by doing a soft-allocation or hard-allocation during the discovery phase, but as the text stands, that is completely up to them. Some vendors might not want to do it.

If Margo's contract states that “Discovery is not a guarantee of reservation,” then Margo must define a standard discovery detail as well. Otherwise, an application that is perfectly compliant with the Margo spec will work flawlessly on Vendor A's hardware but crash randomly on Vendor B's hardware.

Does that map show where the non-deterministic behavior sneaks in?

Reviewer: Ah, ok. I understand what you mean now. 

Isn't this already handled, though, based on how you defined [atomicity](https://github.com/margo/specification-enhancements/blob/feat/infra-allocation-and-conflict-resolution/proposals/profile-discovery-and-resolution/01-capability-framework.md#atomicity)

With the way the desired state works, the device will pull down the latest desired state, which would indicate two applications needing to be deployed, so the device starts deploying them

- With the atomicity rule, the first app would see that port 8080 is free, so the deployment process would continue for that app. Once the deployment for that app is successful, it moves to the next app.
- With the atomicity rule, it would see that port 8080 is no longer free, so the deployment of that app would fail.

Shouldn't the act of having a successful deployment "claim" that resource if it's allocatable?

Maybe I'm making some assumptions that the spec doesn't define as far as apps being deployed sequentualy install of in parallel.

Author: hmm.. that assumption can solve this problem but actually it is nowhere specified in the spec and we never considered sequential guarantee even while writing sandbox. So, I think if we want to mandate sequential installs, then we should add this to the official spec so that everyone be well aware of this.

Reviewer: Yeah, I see your original concern, and these rules probably do need to be defined so devices don't run into the problem you've identified. 

I don't know how prescriptive the rule needs to be. 

It might be possible to have a requirement that indicates when a device is installing/updating multiple applications as part of the same desired state, they need to ensure their chosen deployment approach does not fail the atomicity check because they didn't take into account other applications being installed as part of the same desired state. We may not need to define how they protect against this, just indicate they need to.

---

> ### ProfileDiscoveryRequest — A Materialization of the newly introduced discoverySchema

Reviewer: After reading through the updates, I'm confused by this `discoverProfiles` object. The SUP doesn't fully explain how the information in this section gets from the ApplicationDescription. Also, in some of the examples below, it seems like it's just a repeat of what the device sent to the WFM when it sends up its device profile object.

Author: The `discoverProfiles` is deliberately never a part of the `ApplicationDescription`. The reason for this is to enforce a strict separation of concerns. My take is that we should keep the developer completely out of the discovery business.

The Developer defines the "Contract": All they should do is expose clean, well-documented parameters stating exactly what the application needs to run (e.g., gpu_required: AMD or default_port: 8080). They shouldn't have to think about infrastructure discovery mechanisms at all.

The End-User/WFM handles the "Binding": It should be completely up to the end-user or the WFM to write the profile discovery. They use the it to bind those abstract needs to actual, live hardware at execution time.

If we let developers write profile discovery inside the application description, we create a massive maintenance risk: if we ever make a change to the `ProfileDefinition` in official spec then every single application description using it will go invalid overnight. 

But, by limiting this responsibility strictly to the end-user/WFM at the deployment layer (ApplicationDeployment object), the impact surface area of a spec update becomes much smaller and easier to manage.

Just my take. Pardon if it sounds too naive or assumptive.

Reviewer: I think we're misunderstanding each other.

Let's take this requirement as an example. 

I am an application supplier with an application that requires a GPU with at least 8 GiB of RAM and Ampere architecture. I would like the WFM to be able to match my application requirements to a compatible device.

## Current Specification
In the current specification, we have a limited way for the device supplier to indicate what the device can provide, and a limited way for the application vendor to indicate what their application needs. With the information from both sides, the WFM can attempt some smart pairing of application to device.

- Device supplier [Device Capabilities](https://docs.margo.org/specification/margo-management-interface/device-capabilities) - Through the current device capabilities, the device supplier can indicate their device has a GPU by specifying a peripheral of type `gpu`. However, our current implementation is limited, and the device supplier cannot indicate how much RAM or what architecture the GPU is.
- Application supplier [Application Description](https://docs.margo.org/specification/applications/application-description#requiredresources-attributes) - Through the current application description, the application supplier can indicate their application requires a GPU by setting this in the `RequiredResources` section of the application description. However, since the current `RequiredResources` is a direct map to `Resources` in the device capabilities, the application supplier also cannot indicate how much RAM and what architecture the GPU needs.

## Device Manifest Bisection SUP
The [new SUP](https://github.com/margo/specification-enhancements/blob/sup/bisect-device-manifest/proposals/sup_bisect_device_manifest.md) you're working on re-imagines how device capabilities should be implemented, and it provides for a more dynamic and detailed way for a device to express what it has and what it can do. This SUP is scoped to the device capabilities only and does not propose any changes to the application description (I'm not saying it needs to, just using this to clarify my original point)

- Device supplier device profiles - Through the enhanced profiles SUP, the device supplier now has a way to express much more details about the GPU on their device via the `deviceprofile.margo.org/peripherals/gpu`. So, now, as a device supplier, I have a way to communicate not only that I have a GPU, but it has 24 GiB of vRAM and the "Ampere" architecture.
- Application supplier [Application Description](https://docs.margo.org/specification/applications/application-description#requiredresources-attributes) - There are currently no proposed changes to the application description, so even though the device supplier is capable of communicating that their device has a GPU with 24GiB of vRAM and the "Ampere" architecture, the application supplier is still only limited to saying they need a GPU and not how much RAM or what architecture they need

## Profie Discovery & Resolution
This [additional follow-on SUP](https://github.com/margo/specification-enhancements/blob/feat/infra-allocation-and-conflict-resolution/proposals/profile-discovery-and-resolution/01-capability-framework.md) introduces some new layers on top of the device profiles. It provides a mechanism for how an application vendor can use a combination of the functionality being proposed in the [device-specific parameter value SUP](https://github.com/margo/specification-enhancements/pulls) with the new functionality in this profile discovery SUP. These proposed changes now allow the application vendor to get parameter values that are required (e.g., the devicePath), but it still doesn't address how the application vendor indicates their application requires a GPU with at least 8 GiB of RAM and Ampere architecture

## Original question
So, my original question was intended to see if you had given any thought to how the ApplicationDescription would be updated to also allow the application supplier to indicate more details about what their application's required resources are.

- Do you have any intentions on defining changes to the ApplicationDescription in either of these SUPs, or a follow-on SUP?
- If you've thought about it, do you see it as extending the current application description's [RequiredResources](https://docs.margo.org/specification/applications/application-description#requiredresources-attributes) object to allow specifying more details about the required resources? Or do you see a more significant change also needed in the application description so things are more dynamic for the application supplier as well?
- If you haven't considered this, is this something you would expect someone else to define as a companion to your SUPs?

> A **ProfileDiscoveryRequest** is authored by an end-user persona or wfm inside an `ApplicationDeployment`. It is a materialization of the discovery schema defined in the `ProfileDefinition`. It is a **deferred declaration** — authored before deployment, resolved by the platform at evaluation time (when the deployment is triggered).

Reviewer: The WFM is the only thing that will create the ApplicationDeployment.yaml

```suggestion
A **ProfileDiscoveryRequest** is authored by the wfm inside an `ApplicationDeployment`. It is a materialization of the discovery schema defined in the `ProfileDefinition`. It is a **deferred declaration** — authored before deployment, resolved by the platform at evaluation time (when the deployment is triggered).
```

Author: I'm not sure if the spec actually has any stance on this, but I had an impression that an `ApplicationDeployment` can be authored in multiple ways and one of them could be:

> An end-user comes to a WFM Marketplace, selects an application description and overrides whatever params are allowed, and then triggers a deployment.

Now this is completely upto the WFM implementation how they prepare the `ApplicationDeployment`. It could be prepared completely on a UI and then sent to the WFM, or it can be done completely on the WFM. It can ask user to override params as well. This is why, I added `authored by end-user persona or wfm`, as I am not sure what the implementation could be.

Reviewer: Marketplaces are not in scope for Margo. 

I'll chalk this up to semantics, so I'll drop this comment.

- Performed by the Device Agent
- Evaluated against the device's current `ProfileState`

### Evaluation Timing

1. **Authoring time** — End-user declares `ProfileDiscoveryRequest` with `valueFrom` references
2. **Pre-flight time (WFM)** — WFM validates request against `requestSchema` and may perform advisory checks using last known `ProfileState` as reported by the Device. This is upto the implementation.
3. **Evaluation time (Device Agent)** — Device Agent resolves device-scoped profiles against current `ProfileState`
4. **Injection time** — Resolved values replace `valueFrom` references in parameters
5. **Execution time** — Deployment executes with concrete values

### Atomicity

Resolution of all `discoverProfiles` entries within a single `ApplicationDeployment` is **atomic**: either all entries resolve successfully or the entire deployment is rejected. Partial resolution is not permitted. If any single `ProfileDiscoveryRequest` fails, the platform MUST:

- Reject the deployment before any component is installed
- Return a structured failure identifying which `profileDefinitionId` caused the failure and the applicable failure code

> This aligns with the all-or-nothing batch semantics established in [`sup_bisect_device_manifest`] for `ProfileState` submission.

Reviewer: Was this "[`sup_bisect_device_manifest`]" part intended to be a link as well?

Author: Yup, I'll fix it. Thanks!

- Never return a failure code not declared in the `ProfileDefinition`

---

## Scope Extensibility

This SUP defines only **device-scoped** resolution. The `scope` field in `ProfileDefinition` is intentionally designed to be extensible.

**Example of a future scope — Fleet (WFM-managed):**
Fleet-scoped profiles would represent resources managed at the fleet level rather than on individual devices — for example:

- Secrets backends (e.g., Vault)
- TLS certificate provisioning

In a fleet-scoped model, the WFM would resolve `ProfileDiscoveryRequest` entries against WFM-managed `ProfileState`, and inject resolved outputs into the deployment before dispatching it to the device.

**Why this SUP does not introduce fleet or other scopes:**
- Fleet-scoped resolution requires WFM-side `ProfileState` authoring and lifecycle semantics that are not yet defined
- Introducing fleet scope now would create normative obligations on WFM vendors before the contract is sufficiently specified
- Device-scoped resolution is the foundational building block; fleet scope builds on top of it, and then other options can also be.

---

## Use Cases

### Use Case 1 — GPU Discovery (Generator)

> An ML inference application needs a GPU with at least 8 GiB VRAM and Ampere architecture.

Reviewer: This use case is missing the ApplicationDescription YAML showing how the app vendor indicates it needs an 8GiB VRAM GPU with Ampere. 

It would be helpful if each use case showed the ApplicationDescription,yaml (can be snippets) the app vendor provides, the ApplicationDeployment.yaml the WFM provides, and the ProfileDefinition.

Reviewer: See [this detailed comment](https://github.com/margo/specification-enhancements/pull/66#discussion_r3645778997) about why I'm asking about this.

    request:
      minVramGiB: 8
      architecture: ampere

parameters:
  gpuDevicePath:
    valueFrom: discoverProfiles.gpuForInference.output.devicePath
    targets:
      - pointer: GPU_DEVICE_PATH
        components: ["ml-inference"]
```

**Resolution outcome:** Device Agent finds a matching GPU, resolves `devicePath`, and injects it into the `GPU_DEVICE_PATH` parameter before the application starts.

---

### Use Case 2 — Port Availability (Guard)

A web server application needs port `8080` to be available on the device before deployment.

**ProfileDefinition:**
```json
{
  "profileDefinitionId": "deviceprofile.margo.org/network/port",
  "scope": "device",
  "category": "resource",
  "platformStateSchema": {...},
  "discoverySchema": {
    "mode": "guard",
    "requestSchema": {
      "type": "object",
      "properties": {
        "port": { "type": "number" },
        "protocol": { "type": "string" }
      },
      "required": ["port"]
    },
    "outputSchema": {},
    "failureCodes": ["PortUnavailable", "PortReserved"]
  }
}
```

**ApplicationDeployment:**
```yaml
discoverProfiles:
  httpPort:
    id: deviceprofile.margo.org/network/port
    request:
      port: 8080

>       protocol: tcp

Reviewer: This is kind of confusing. I think I need to see the ApplicationDescription.yaml to understand what is happening. The fact that an app wants to use TCP on Port 8080 needs to come from the ApplicationDescription, so I'd like to see how you envision this information being provided.

Reviewer: See [this detailed comment](https://github.com/margo/specification-enhancements/pull/66#discussion_r3645778997) about why I'm asking about this.

```

**Resolution outcome:** Device Agent checks whether port `8080` is available. If occupied, it returns `PortUnavailable` and rejects the deployment before any component is installed.

---

### Use Case 3 — Available Network Port (Generator)

An application needs any available port assigned dynamically rather than specifying one explicitly.

**ProfileDefinition:**
```json
{

>   "profileDefinitionId": "deviceprofile.margo.org/network/dynamic-port",

Reviewer: I'm not a fan of devices having to provide both `deviceprofile.margo.org/network/port` and `deviceprofile.margo.org/network/dynamic-port` profiles. There's too much duplication.

I think it would be better if a profile could handle both cases instead.

Reviewer: @singhmj-1 what do you think about the idea of adding a "validatedBy" property to parameters? This would allow for a profile definition to be either a guard, generator, or both

Here are a couple of examples of what I mean. 

### Network Ports
Instead of having `deviceprofile.margo.org/interfaces/port` guard and `deviceprofile.margo.org/interfaces/dynamic-port` generator, there is just  `deviceprofile.margo.org/interfaces/port`. An application vendor has two options: they can either indicate they want the port to be automatically assigned, or they can say they want the customer to define the port, but they want the value validated before trying to use it.

#### ProfileDefintion:

Based on your dynamic port example

```jason
{
  "profileDefinitionId": "deviceprofile.margo.org/interfaces/port",
  "scope": "device",
  "category": "resource",
  "platformStateSchema": {...},
  "discoverySchema": {
    "mode": ["generator", "guard"]
    "requestSchema": {
      "type": "object",
      "properties": {
        "protocol": { "type": "string" },
        "rangeMin": { "type": "number" },
        "rangeMax": { "type": "number" }
      }
    },
    "outputSchema": {
      "type": "object",
      "properties": {
        "availablePort": { "type": "number" }
      }
    },
    "failureCodes": ["NoPortAvailable"]
  }
}
```

App vendors can define parameters in two ways:

#### ApplicationDescription Parameter - device-assigned port

The application vendor uses the [proposed](https://github.com/margo/specification-enhancements/blob/device-specific-parameter-values/proposals/device-specific-parameter-values.md) "valueFrom" object to indicate they want the parameter value to be provided by the device's `deviceprofile.margo.org/interfaces/port` availablePort output property (generator)

```yaml
parameters:
  servicePort:
    valueFrom:
      device:
        key: deviceprofile.margo.org/interfaces/port/output/availablePort
        required: true
    targets:
      - pointer: global.service.port
        components:
          - fi-web-helm
      - pointer: SERVICE_PORT
        components:
          - fi-web-compose
```

#### ApplicationDescription Parameter - device port validation only

We introduce a new "validatedBy" object to the ApplicationDescription parameters, allowing the application vendors to indicate they want a property to be validated by a device profile (e.g., the `deviceprofile.margo.org/interfaces/port` profile (guard)).

`validatedBy` (or some other name) is optional and allows an app vendor to say how they want the device to validate a parameter if the parameter value needs to be validated by the device (e.g., network port assignment, ingress routes, etc.)

```yaml
parameters:
  servicePort:
    value: 9095
    validatedBy:
      device:
        key: deviceprofile.margo.org/interfaces/port
    targets:
      - pointer: global.service.port
        components:
          - fi-web-helm
      - pointer: SERVICE_PORT
        components:
          - fi-web-compose
configuration:
  sections:
    - name: Network Configuration
      settings:
       - parameter: servicePort
         name: Service Port
         description: Network port to expose the service on
         schema: rangedNumber
   schema: ...
```

### Ingress Route validation

This would allow for the ingress validation case that was part of the original SUP (and original issue)

#### ProfileDefintion

Just a basic example to demonstrate

```json
{
  "profileDefinitionId": "deviceprofile.margo.org/capability/ingress",
  "description": "Profile Definition for Ingress Route configuration",
  "scope": "device",
  "category": "capability",
  "schemaVersion": "1.0.0",
  "platformStateSchema": {...},
   "discoverySchema": {
     "mode": ["guard"]
   },
  "failureCodes": ["RouteUnavailable"]
}
```

#### ApplicationDescription - ingress route validation only

```yaml
parameters:
  ingressPrefix:
    value: northstar
    validatedBy:
      device:
        key: deviceprofile.margo.org/capability/ingress
    targets:
      - pointer: global.ingress.prefix
        components:
          - fi-web-helm
configuration:
  sections:
    - name: Network Configuration
      settings:
       - parameter: ingressPrefix
         name: Ingress Prefix
         description: The prefix to prepend to the hostname to enable access to the application 
         schema: requiredText
   schema: ...
```

  "scope": "device",
  "category": "resource",
  "platformStateSchema": {...},
  "discoverySchema": {
    "mode": "generator",
    "requestSchema": {
      "type": "object",
      "properties": {
        "protocol": { "type": "string" },
        "rangeMin": { "type": "number" },
        "rangeMax": { "type": "number" }
      }
    },
    "outputSchema": {
      "type": "object",
      "properties": {
        "availablePort": { "type": "number" }
      }
    },
    "failureCodes": ["NoPortAvailable"]
  }
}
```

**ApplicationDeployment:**
```yaml
discoverProfiles:
  servicePort:
    id: deviceprofile.margo.org/network/dynamic-port
    request:
      protocol: tcp
      rangeMin: 8000

>       rangeMax: 9000

Reviewer: This is not information the WFM needs to tell the device because the device already knows this. 

The device sends its `deviceprofile.margo.org/network/dynamic-port` to the WFM, so it doesn't need to send it back to the device.

### Why is resolution atomic?

Partial resolution creates ambiguous deployment states. If two out of three `ProfileDiscoveryRequest` entries resolve but the third fails, the platform cannot safely proceed — some parameters would be injected, others would not. Atomic rejection ensures the deployment either starts with all values resolved or does not start at all.

### Why are `valueFrom` nested paths not supported in this version?

Nested path resolution introduces complexity in schema validation and output traversal. Flat top-level field references cover the majority of practical use cases. Nested paths can be introduced in a future version once the base mechanism is proven.

### Why is `mode` a required field when `discoverySchema` is present?

Without `mode`, the platform must infer evaluation behaviour from schema shape — an unreliable heuristic. Explicit `mode` declaration makes platform behaviour deterministic and removes ambiguity for implementors.

---

## Open Questions

| # | Question | Status |
|---|---|---|

> | 1 | Should `requestSchema` support `required` fields, or are all request inputs always optional? | Open |

Reviewer: Since each device profile has its own schema definition, this can be determined independently.

> | 2 | Should `failureCodes` support structured metadata (e.g., human-readable descriptions) in addition to machine-readable codes? | Open |

Reviewer: I'm ok with just the failure code; that way, the WFM is free to construct its error messages in a way that is consistent with the rest of their site.

A side note on this is globalization. Right now, everything in Margo only supports a single language. At some point, we're going to have to figure out how to support multiple languages. Having human-readable messages here would just add additional overhead to be able to provide these error messages in a bunch of different languages.

> | 3 | Should `valueFrom` support a default value fallback if resolution produces no output for an optional field? | Open |

Reviewer: I think it should be possible to define a default value, but the behavior is the opposite. The default value is used first if it can be used. For example, I may want the port to default to 1234 if it's available; otherwise, another port is assigned.
